druid-handler/src/java/org/apache/hadoop/hive/druid/serde/DruidSerDe.java

```diff
@@ -0,0 +1,379 @@
+    }
+    // Post-aggregator columns
+    for (PostAggregator pa : query.getPostAggregatorSpecs()) {
+      columnNames.add(pa.getName());
+      columnTypes.add(TypeInfoFactory.floatTypeInfo);
+    }
+  }
+
+  /* Select query */
+  private void inferSchema(SelectQuery query, List<String> columnNames,
+          List<PrimitiveTypeInfo> columnTypes) {
+    // Timestamp column
+    columnNames.add(Column.TIME_COLUMN_NAME);
+    columnTypes.add(TypeInfoFactory.timestampTypeInfo);
+    // Dimension columns
+    for (DimensionSpec ds : query.getDimensions()) {
+      columnNames.add(ds.getOutputName());
+      columnTypes.add(TypeInfoFactory.stringTypeInfo);
+    }
+    // Metric columns
+    for (String metric : query.getMetrics()) {
+      columnNames.add(metric);
+      columnTypes.add(TypeInfoFactory.floatTypeInfo);
+    }
+  }
+
+  /* GroupBy query */
+  private void inferSchema(GroupByQuery query, List<String> columnNames, List<PrimitiveTypeInfo> columnTypes) {
+    // Timestamp column
+    columnNames.add(Column.TIME_COLUMN_NAME);
+    columnTypes.add(TypeInfoFactory.timestampTypeInfo);
+    // Dimension columns
+    for (DimensionSpec ds : query.getDimensions()) {
+      columnNames.add(ds.getOutputName());
+      columnTypes.add(TypeInfoFactory.stringTypeInfo);
+    }
+    // Aggregator columns
+    for (AggregatorFactory af : query.getAggregatorSpecs()) {
+      columnNames.add(af.getName());
+      columnTypes.add(DruidSerDeUtils.convertDruidToHiveType(af.getTypeName()));
+    }
+    // Post-aggregator columns
+    for (PostAggregator pa : query.getPostAggregatorSpecs()) {
+      columnNames.add(pa.getName());
+      columnTypes.add(TypeInfoFactory.floatTypeInfo);
+    }
+  }
+
+  @Override
+  public Class<? extends Writable> getSerializedClass() {
+    return NullWritable.class;
+  }
+
+  @Override
+  public Writable serialize(Object o, ObjectInspector objectInspector) throws SerDeException {
+    return NullWritable.get();
+  }
+
+  @Override
+  public SerDeStats getSerDeStats() {
+    throw new UnsupportedOperationException("SerdeStats not supported.");
+  }
+
+  @Override
+  public Object deserialize(Writable writable) throws SerDeException {
+    DruidWritable input = (DruidWritable) writable;
+    List<Object> output = Lists.newArrayListWithExpectedSize(columns.length);
+    for (int i = 0; i < columns.length; i++) {
+      final Object value = input.getValue().get(columns[i]);
+      if (value == null) {
+        output.add(null);
+        continue;
+      }
+      switch (types[i].getPrimitiveCategory()) {
+        case TIMESTAMP:
+          if (value instanceof Long) {
+            output.add(new TimestampWritable(new Timestamp((Long)value)));
+          } else {
+            throw new SerDeException("Cannot transform to TIMESTAMP from Druid type: " +
+                    value.getClass().toString());
+          }
+          break;
+        case LONG:
+          if (value instanceof Byte) {
```

Comment:
The actual type of the value object returned for each row might defer e.g. a long metric might be represented in some rows as a byte and in some others as an int, depending on the value length.

In any case, I realized that we can avoid the if-else using 'Number'. I have changed the code accordingly.